aiosip/peers.py

```diff
@@ -268,27 +266,19 @@ def _create_peer(self, peer_addr):
         return await self._dispatch(protocol, peer_addr)
 
     def connection_lost(self, protocol):
-        peer_addr = protocol.transport.get_extra_info('peername')
-        local_addr = protocol.transport.get_extra_info('sockname')
-        peer = self._peers.pop((peer_addr, local_addr), None)
-        if peer:
-            peer.close()
+        for key, peer in list(self._peers.items()):
+            if peer._protocol == protocol:
+                peer._disconnected()
+                self._peers.pop(key)
 
     async def close(self):
-        for peer in self._peers.values():
-            peer.close()
+        for peer in list(self._peers.values()):
+            await peer.close()
 
         for server in self._servers.values():
             server.close()
             await server.wait_closed()
 
```

Comment:
We probably also want to empty `self._protocols`, `self._servers` and `self._peers` here.

Reply:
self._peers (and self._protocols once I address your previous comment) will already get emptied by connect_lost().

You're right about the servers though. This however brings me to another point : in this case we need to prevent users from shutting down a server "manually" but calling .close() on what Application.run() returns. Maybe we could simply make Application.run() not return anything?